### ISSUE TITLE
Continous Exceptions when uploading large amount of documents in a loop

### ISSUE THREAD

**Issue by dmm-l-mediehus** [kbeaugrand/SemanticKernel.Connectors.Memory.SqlServer#92]
Look at the middle part of this warning:
`at KernelMemory.MemoryStorage.SqlServer.SqlServerMemory.UpsertAsync...`:
```
warn: Microsoft.KernelMemory.Orchestration.AzureQueues.AzureQueuesPipeline[0]
      Message 'e88b5c65-3033-429f-954d-6332904b6f94' processing failed with exception, putting message back in the queue with a delay of 5000 msecs
      Microsoft.Data.SqlClient.SqlException (0x80131904): Execution Timeout Expired.  The timeout period elapsed prior to completion of the operation or the server is not responding.
       ---> System.ComponentModel.Win32Exception (258): The wait operation timed out.
         at Microsoft.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction)
         at Microsoft.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose)
         at Microsoft.Data.SqlClient.SqlCommand.InternalEndExecuteNonQuery(IAsyncResult asyncResult, Boolean isInternal, String endMethod)
         at Microsoft.Data.SqlClient.SqlCommand.EndExecuteNonQueryInternal(IAsyncResult asyncResult)
         at Microsoft.Data.SqlClient.SqlCommand.EndExecuteNonQueryAsync(IAsyncResult asyncResult)
         at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization)
      --- End of stack trace from previous location ---
         at KernelMemory.MemoryStorage.SqlServer.SqlServerMemory.UpsertAsync(String index, MemoryRecord record, CancellationToken cancellationToken)
         at Microsoft.KernelMemory.Handlers.SaveRecordsHandler.SaveEmbeddingsAsync(DataPipeline pipeline, CancellationToken cancellationToken)
         at Microsoft.KernelMemory.Handlers.SaveRecordsHandler.InvokeAsync(DataPipeline pipeline, CancellationToken cancellationToken)
         at Microsoft.KernelMemory.Pipeline.DistributedPipelineOrchestrator.RunPipelineStepAsync(DataPipeline pipeline, IPipelineStepHandler handler, CancellationToken cancellationToken)
         at Microsoft.KernelMemory.Pipeline.DistributedPipelineOrchestrator.<>c__DisplayClass3_0.<<AddHandlerAsync>b__0>d.MoveNext()
      --- End of stack trace from previous location ---
         at Microsoft.KernelMemory.Orchestration.AzureQueues.AzureQueuesPipeline.<>c__DisplayClass24_0.<<OnDequeue>b__0>d.MoveNext()
      ClientConnectionId:56331e50-c92e-4bb9-89b8-803b64382545
      Error Number:-2,State:0,Class:11
```

My KernelMemory orchestration (AzureQueues) calls your UpsertAsync which causes different exceptions on the SQL database:
`System.InvalidOperationException: Timeout expired.  The timeout period elapsed prior to obtaining a connection from the pool.  This may have occurred because all pooled connections were in use and max pool size was reached.
         at Microsoft.Data.Common.ADP.ExceptionWithStackTrace(Exception e)`
and
`Microsoft.Data.SqlClient.SqlException (0x80131904): Execution Timeout Expired.  The timeout period elapsed prior to completion of the operation or the server is not responding.
       ---> System.ComponentModel.Win32Exception (258): The wait operation timed out.`

I'm uploading about 60,000 documents from my SQL database into the KernelMemory:
```cs
...looping documents...

await kernelMemory.ImportTextAsync(
	text: documentSemanticText.ToString(),
	documentId: document.DocumentID.ToString(),
	index: "documents",
	cancellationToken: stoppingToken);

...end of loop
```

Configured this way:
```cs
public static IServiceCollection AddKernelMemory(this IServiceCollection services, IConfiguration configuration)
{
	KernelMemoryBuilder memoryBuilder = new(services);

	KernelMemoryConfig.Configuration = configuration;

	memoryBuilder.WithOpenAITextGeneration(KernelMemoryConfig.TextGenerationOpenAIConfig);
	memoryBuilder.WithOpenAITextEmbeddingGeneration(KernelMemoryConfig.EmbeddingGenerationOpenAIConfig);
	memoryBuilder.WithAzureQueuesOrchestration(KernelMemoryConfig.AzureQueuesConfig);
	memoryBuilder.WithSqlServerMemoryDb(KernelMemoryConfig.AIConnectionString);
	memoryBuilder.WithSearchClientConfig(KernelMemoryConfig.SearchClientConfig);
	memoryBuilder.WithCustomTextPartitioningOptions(KernelMemoryConfig.TextPartitioningOptions);

	IKernelMemory kernelMemory = memoryBuilder.Build();

	services.AddSingleton(kernelMemory);

	return services;
}
```

my own `KernelMemoryConfig`:
```cs
public static class KernelMemoryConfig
{
	public static IConfiguration Configuration { get; set; } = null!;

	public static string AIConnectionString => Configuration.GetConnectionString("AIDataContext")!;

	public static OpenAIConfig TextGenerationOpenAIConfig => new()
	{
		APIKey = Configuration["OpenAI:Settings:APIKey"]!,
		OrgId = Configuration["OpenAI:Settings:OrgId"]!,
		TextModel = Configuration["OpenAI:Models:GPT"]!
	};

	public static OpenAIConfig EmbeddingGenerationOpenAIConfig => new()
	{
		APIKey = Configuration["OpenAI:Settings:APIKey"]!,
		OrgId = Configuration["OpenAI:Settings:OrgId"]!,
		EmbeddingModel = Configuration["OpenAI:Models:Embedding"]!
	};

	public static AzureQueuesConfig AzureQueuesConfig => new()
	{
		ConnectionString = Configuration["Azure:Queue:ConnectionString"]!,
		AccountKey = Configuration["Azure:Queue:AccountKey"]!,
		Account = Configuration["Azure:Queue:Account"]!,
		Auth = AzureQueuesConfig.AuthTypes.ConnectionString
	};

	public static SearchClientConfig SearchClientConfig => new()
	{
		EmptyAnswer = "Jeg beklager, men jeg har ikke fundet nogen relevant information om det, der kan bruges til at besvare dit spørgsmål.",
		MaxMatchesCount = 10,
		AnswerTokens = 800
	};

	public static TextPartitioningOptions TextPartitioningOptions => new()
	{
		MaxTokensPerParagraph = 1000,
		MaxTokensPerLine = 300,
		OverlappingTokens = 100
	};
}
```

It still creates the embeddings, but it's making the UI application unresponsive (spinning/loading) during the import of the documents, despite using the Service (not Serverless)...

After swapping out with .WithAzureAISearchMemoryDb(...) instead of the KernelMemory SqlServer - it works as expected (fast UI), normal imports, fast SearchAsync, etc. So something is wrong.

**Comment by kbeaugrand** [comment]
@dmm-l-mediehus this might be caused since importing document are made in parallel. 
There might be as much SQL connections opened as thread that are working at a time. Since the SQL connections are pooled you might rapidly reach the maximum number of connection pooled allowed in the SQL pool.

You'll have 2 options : 

- Without changing anything in your code, you can increase the ```Max Pool Size``` setting in your connection string. The default is 100 but you can increase it a little bit in case you don't want to impact your code. 
- While changing your code, you can add in your loop a verification of document readiness before going to the next one...

ex: 
```
...looping documents...

await kernelMemory.ImportTextAsync(
	text: documentSemanticText.ToString(),
	documentId: document.DocumentID.ToString(),
	index: "documents",
	cancellationToken: stoppingToken);

while (true)
  if (await kernelMemory.IsDocumentReadyAsync(document.DocumentID.ToString(), "documents", stoppingToken))
    break;

  await Task.Delay(TimeSpan.FromMilliseconds(200), stoppingToken);

...end of loop
```

**This is generally not a good option to increase the max pool size. I strongly advise you to reduce the degree of parallelism in your documents import job to limit the amount of resources (in your batch job and also for your SQL server instance) needed.**

**Comment by dmm-l-mediehus** [comment]
Are you going to change anything in your nuget package about this issue in the future?

Changing max pool size I have never done in my many years of programming with SQL servers. So I agree it might not be good to do so.

Also I **don't** think that importing ~60k items waiting for each item to upload, before going for the next one, is good.

I would like to reduce the parallelism when I upload using BackgroundService, so it doesn't take up everything, but not sure how I would go about this. At the moment I call my database and materialize 60k items in a List<>, then ImportTextAsync with KernelMemory. Do you have any recommended ideas?

**Comment by kbeaugrand** [comment]
@dmm-l-mediehus, Let me some time to investigate on that please, I'll discuss with the KM team to understand more deeply how the KM works when looping on imports the limits and so...

In the same time, I'll checked also the connector to be sure that every connection is released correctly when doing operations, to clarify that point...

**Comment by dmm-l-mediehus** [comment]
Tried making it 200 max pool size, didn't work still.

Then I waited for document to upload & wait 1 sec... before going to next document. And that worked. That'll do for now.

**Comment by kbeaugrand** [comment]
> Tried making it 200 max pool size, didn't work still.
> 
> Then I waited for document to upload & wait 1 sec... before going to next document. And that worked. That'll do for now.

When waiting for document uploads, can you take a look at active connexions on sql server to see if the connexion pool is stable ?

**Comment by dmm-l-mediehus** [comment]
Oh nevermind, the wait for document to upload & wait 1 sec .. is still causing the error.

```
Execution Timeout Expired. The timeout period elapsed prior to completion of the operation or the server is not responding. | The wait operation timed out. | at Microsoft.Data.SqlClient.SqlConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction) at Microsoft.Data.SqlClient.SqlInternalConnection.OnError(SqlException exception, Boolean breakConnection, Action`1 wrapCloseInAction) at Microsoft.Data.SqlClient.TdsParser.ThrowExceptionAndWarning(TdsParserStateObject stateObj, Boolean callerHasConnectionLock, Boolean asyncClose) at Microsoft.Data.SqlClient.SqlCommand.EndExecuteReaderInternal(IAsyncResult asyncResult) at Microsoft.Data.SqlClient.SqlCommand.EndExecuteReaderAsync(IAsyncResult asyncResult) at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization) --- End of stack trace from previous location --- at KernelMemory.MemoryStorage.SqlServer.SqlServerMemory.GetIndexesAsync(CancellationToken cancellationToken) at KernelMemory.MemoryStorage.SqlServer.SqlServerMemory.DoesIndexExistsAsync(String indexName, CancellationToken cancellationToken) at KernelMemory.MemoryStorage.SqlServer.SqlServerMemory.GetSimilarListAsync(String index, String text, ICollection`1 filters, Double minRelevance, Int32 limit, Boolean withEmbeddings, CancellationToken cancellationToken)+MoveNext() at KernelMemory.MemoryStorage.SqlServer.SqlServerMemory.GetSimilarListAsync(String index, String text, ICollection`1 filters, Double minRelevance, Int32 limit, Boolean withEmbeddings, CancellationToken cancellationToken)+System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult() at Microsoft.KernelMemory.Search.SearchClient.SearchAsync(String index, String query, ICollection`1 filters, Double minRelevance, Int32 limit, CancellationToken cancellationToken) at Microsoft.KernelMemory.Search.SearchClient.SearchAsync(String index, String query, ICollection`1 filters, Double minRelevance, Int32 limit, CancellationToken cancellationToken) at Landbrugnet.AI.Blazor.Pages.AIService.SearchAsync() in G:\LandbrugnetAI\Landbrugnet.AI.Blazor\Pages\AIService.razor:line 160
```
Code:
```cs
await kernelMemory.ImportTextAsync(
	text: documentSemanticText.ToString(),
	documentId: document.DocumentID.ToString(),
	index: "documents",
	cancellationToken: stoppingToken);

while (!await kernelMemory.IsDocumentReadyAsync(document.DocumentID.ToString(), "documents", stoppingToken))
{
	await Task.Delay(1000, stoppingToken);
}
```

I did [sp_who](http://msdn.microsoft.com/en-us/library/ms174313.aspx) before, and it says 1 connection on the sql database.. So idk what's going on.

**Comment by dmm-l-mediehus** [comment]
Nice changes on the commit, looks like a fix to me, let me know when it's out for download on nuget :D

**Comment by kbeaugrand** [comment]
Yes @dluc helped me find where the problem was coming from.

I'm about to finish, the nuget will arrive in a few minutes.

**Comment by kbeaugrand** [comment]
@dmm-l-mediehus nuget package for 1.1.3 is now available with this fix. Feel free to reach me out in case you encounter new issues...

**Comment by dmm-l-mediehus** [comment]
There are no longer any warnings.

But after I removed the waiting for document to upload code (+1 sec wait), this is the behavior:

The UI is still blocked despite using BackgroundService (WorkerService) for uploading the data (chrome loading spinning on tab). This doesn't happen using AzureAI's nuget package for memory storage.

After some time of uploading (~3 minutes) - both uploading and web UI is disconnects completely. Then a few minutes later it comes back to live for a minute where I can do stuff just fine, but then a minute after; it blocks again for another few minutes..And disconnects. Repeatedly.
![image](https://github.com/kbeaugrand/SemanticKernel.Connectors.Memory.SqlServer/assets/125255354/50a945a8-4082-4e3a-9198-f6c5691ed12a)
 
As if there's blockage up until a point where something disposes, then it works fine, and it repeats the blockage, etc.

**Comment by kbeaugrand** [comment]
@dmm-l-mediehus? can you open a new issue regarding this problem? I'll take a look on it.
If you can also provide me a sample code demonstrating this issue, it could be very usefull

**Comment by dmm-l-mediehus** [comment]
Issue created here: https://github.com/kbeaugrand/SemanticKernel.Connectors.Memory.SqlServer/issues/95
